src/gsy_e/models/strategy/heat_pump.py

```diff
@@ -0,0 +1,152 @@
+from typing import Dict, TYPE_CHECKING
+
+from gsy_framework.constants_limits import ConstSettings
+from gsy_framework.data_classes import Trade
+from gsy_framework.enums import AvailableMarketTypes
+from pendulum import DateTime, duration
+
+from gsy_e.constants import FLOATING_POINT_TOLERANCE
+from gsy_e.models.state import HeatPumpState
+from gsy_e.models.strategy.energy_parameters.heat_pump import HeatPumpEnergyParameters
+from gsy_e.models.strategy.trading_strategy_base import TradingStrategyBase
+from gsy_e.models.strategy.order_updater import OrderUpdaterParameters, OrderUpdater
+
+DEFAULT_HEAT_PUMP_ORDER_UPDATE_PARAMS = {
+    AvailableMarketTypes.SPOT: OrderUpdaterParameters(
+        initial_rate=ConstSettings.HeatPumpSettings.BUYING_RATE_RANGE.initial,
+        final_rate=ConstSettings.HeatPumpSettings.BUYING_RATE_RANGE.final,
```

Comment:
@spyrostz We have a requirement, that by default (if the user does not provide one):
- the final rate should be the MMR by default
- the initial rate should be the FIT by default 
But now I have a problem. The `InfiniteBusStrategy` could be initiated after this one and these global settings could not be updated in time for the init of the HP class. Also, if the MMR is variable, this is even more complicated.
I propose to allow None values in the `OrderUpdaterParameters` data class and deal with this in the `_create_order_updaters` class with something like:
```python
if not self._order_updater_params[market_type].final_rate:
   self._order_updater_params[market_type].final_rate = GlobalConfig.market_maker_rate[market_slot]
```
Or do you have a better idea?

Reply:
I agree, however you already have default values for MMR and FIT, therefore I do not see the need to pass None values to the OrderUpdaterParameters class. This also covers the bug / cornercase where neither user parameters, nor MMR / FIT are passed to this strategy. This will make it flexible enough to work without any parameters at all IMO, no matter if we decide to force these parameters in the gsy-web validation or not.

Reply:
@spyrostz 
I am confused by the first part of your comment unfortunately.
> ... therefore I do not see the need to pass None values to the OrderUpdaterParameters class. ...

But I need to distinguish between the case when the user actually sets the initial and final rate and the case when they want to use the current FIT and MMR. Or am I missing something here?

Reply:
Distinguishing between these 2 cases does not need to be part of the OrderUpdaterParameters class though, and IMO it should be part of the strategy (by e.g. having a boolean variable that dictates which of the 2 cases is active, the MMR/FIT or the user-defined rate). Reason that I believe that this would be preferable, is because the OrderUpdaterParameters is a dataclass that encapsulates only the parameters of the order updater, and ideally should not contain any strategy-specific logic, since this will affect other strategies too. For instance, if you add such logic to the OrderUpdaterParameters, then this logic would be imported to the forward strategies. Even if it is not used by the forward markets, it is probably better to not be part of the class in the first place.

Reply:
Ah yes, did not want to add it to the `OrderUpdaterParameters`, but to the Strategy. However, members of the `OrderUpdaterParameters` should be able to be set to None AFAIU, in order to set the `DEFAULT_HEAT_PUMP_ORDER_UPDATE_PARAMS` so that I can distinguish in the strategy if the user set them or not. Does it make sense?

Reply:
I think this is the only minor point where we do not agree in our comments. IMO allowing the members of the OrderUpdaterParameters to be None would be implicit behavior, because you would then have to check one of the members whether it is None or not. The explicit behavior would be to have a separate boolean variable in the strategy class (e.g. _are_user_parameters_set) that can be used in a more clear way. I have 2 arguments in favor of the latter:
- it is more misleading / confusing for the unaware reader to see something like this `if self._energy_params.initial_rate is None: <use parameters from FIT/MMR>` compared to this `if not self._are_user_parameters_set: <use parameters from FIT/MMR>`. 
- in reality, in order for the OrderUpdaterParameters to be a usable object, they have to be populated, and in order to protect ourselves from the None case we would have to add if / else blocks in actions that depend on OrderUpdaterParameters. For now the only place that we have to do this is the _create_order_updaters, however if we add more methods that use the OrderUpdaterParameters this will quickly become convoluted IMO

Reply:
Thank you. The `_are_user_parameters_set` parameter makes sense. However, the user is limited to two choices:
1. Set all `OrderUpdaterParameters` that are static for the whole simulation 
2. Not set `OrderUpdaterParameters` and let the strategy use the default ones (which will be initial_rate=FIT; final_rate=MMR, update_interval=DEFAULT_UPDATE_INTERVAL)
But there is also the following case, when the user wants to set e.g. only the final_rate to a constant value.
Do you think we should support this case? We can also resume this discussion offline (and post an outcome here) as the number of comments has already exceeded our workflow limit.